### ISSUE TITLE
PropertyGrid: Prevent invalid data from being assigned to the property

### ISSUE THREAD

**Issue by aldinei-sampaio** [objorke/PropertyTools#61]
I have some validations I need to do in my properties, like:
- Required (should not be null or empty string)
- Regex (entered text must follow a predefined regex pattern)
- Range (max and min values for numeric properties)

I successfully managed to configure my class as the PropertyGrid validates and displays error messages for all of those conditions. However, I had not found a way to prevent the PropertyGrid from assigning the invalid values to the property.

Example: In the "DataAnnotations" example of PropertyGridDemo, there is a property nammed "RequiredString". If I delete the contents of the textbox, the PropertyGrid shows the error message, but the empty string was already assigned to the property. That's easy to reproduce as follows:

In the TestBase class, add this method:

        protected void RaisePropertyChanged(string propertyName)
        {
            if (PropertyChanged != null)
            {
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
            }
        }

And in the TestDataAnnotations class, change this code:

        [Required(AllowEmptyStrings = false, ErrorMessage = "A value is required.")]

For this:

        private string _requiredString;
        [Required(AllowEmptyStrings = false, ErrorMessage = "A value is required.")]
        public string RequiredString
        {
            get
            {
                return _requiredString;
            }
            set
            {
                if (value != _requiredString)
                {
                    _requiredString = value;
                    RaisePropertyChanged("RequiredString");
                    RaisePropertyChanged("CurrentRequiredString");
                }
            }
        }

        public string CurrentRequiredString
        {
            get
            {
                return _requiredString;
            }
        }

Now, when the user clears the contents of "RequiredString" textbox, the "CurrentRequiredString" textbox is also cleared.

**Comment by objorke** [comment]
Isn´t the current behavior consistent with the standard behavior for validations based on `System.ComponentModel.DataAnnotations`? 
Isn´t it better to handle this on the model side than in the `PropertyGrid`?

**Comment by aldinei-sampaio** [comment]
Yes, you are right. After some research, I see that the behavior is actually exactly like the DataAnnotations pattern (that I was unaware of before).

I managed to ensure my business rules on the model. Throwing an exception on the property setter created exactly the behavior I needed.

However, I noticed that the ValidationErrorTemplate is not used to show messages from databinding errors. The textbox border turns red and the validation error appears on the tooltip, but the ValidationErrorTemplate is not rendered. Seems that ValidationErrorTemplate is rendered only for DataAnnotation validation errors.

That can be seem in the PropertyGrid Demo, on the "Validation with IDataErrorInfo" option. In the "Age" field, if I enter the text "ABC" the behavior of the PropertyGrid is not the same as if I enter the text "-1".
